### DISCUSSION TITLE
Add support for secrets in `__dangerous__body_down_if_text_missing`

### DISCUSSION THREAD

**Discussion by KaKi87** [upptime/upptime#898] [Ideas] [answered]
**Describe the bug**
Using an [Actions secret](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions#creating-secrets-for-a-repository) in `__dangerous__body_down_if_text_missing` doesn't work.

**To Reproduce**
1. Create a `PRIVATE_SERVICE_DOWNIFMISSING` Actions secret ;
2. Add `__dangerous__body_down_if_text_missing: $PRIVATE_SERVICE_DOWNIFMISSING` to `.upptimerc.yml`.

**Expected behavior**
Upptime checks whether the secret text is missing.

**Screenshots**
https://github.com/server-home-KaKi87/status/blob/df013dc0b0d852b6d3403cf3d058d76a36dc294d/.upptimerc.yml#L14

**Desktop (please complete the following information):**
 - OS: Ubuntu Unity
 - Browser: Brave
 - Version 1.60.125

**Smartphone (please complete the following information):**
 - Device: Xiaomi Mi 8
 - OS: Android
 - Browser Kiwi
 - Version 120.0.6099.26

**Additional context**
I wonder why does this issue template requires desktop & mobile devices info.

---
Thanks

**Comment by AnandChowdhary** [comment]
This is how we currently implement the check: https://github.com/upptime/cli/blob/c6e6c3b24bc0a2a92c13bc3a08b898a2a1075260/src/update.ts#L133

Any ideas what could be going wrong?

**Comment by KaKi87** [comment]
Well, that works for normal strings, but not for Actions secrets.

Could you point me to the implementation of `url` and `headers`, which are documented as supporting Actions secrets ?

Thanks

**Comment by AnandChowdhary** [comment]
This is how we deal with Actions secrets in `url` - https://github.com/upptime/uptime-monitor/blob/master/src/update.ts#L158

This is the source for `replaceEnvironmentVariables` - https://github.com/upptime/uptime-monitor/blob/master/src/helpers/environment.ts

I don't see it anywhere that you can use an Actions secret (environment variable) for `__dangerous__body_down_if_text_missing`. Is this somewhere documented?

**Comment by KaKi87** [comment, marked as answer]
> This is how we deal with Actions secrets in `url` - [upptime/uptime-monitor@`master`/src/update.ts#L158](https://github.com/upptime/uptime-monitor/blob/master/src/update.ts?rgh-link-date=2023-12-15T14%3A06%3A19Z#L158)

Thanks, then I suggest replacing these :

```js
result.data.includes(site.__dangerous__body_down)
result.data.includes(site.__dangerous__body_degraded)
!result.data.includes(site.__dangerous__body_degraded_if_text_missing)
!result.data.includes(site.__dangerous__body_down_if_text_missing)
```

With these :

```js
result.data.includes(replaceEnvironmentVariables(site.__dangerous__body_down))
result.data.includes(replaceEnvironmentVariables(site.__dangerous__body_degraded))
!result.data.includes(replaceEnvironmentVariables(site.__dangerous__body_degraded_if_text_missing))
!result.data.includes(replaceEnvironmentVariables(site.__dangerous__body_down_if_text_missing))
```

> I don't see it anywhere that you can use an Actions secret (environment variable) for `__dangerous__body_down_if_text_missing`. Is this somewhere documented?

It's not, I just hoped it would work, and when it didn't I submitted this post.

Thanks

**Reply by AnandChowdhary** [reply]
That's a good idea! Would you like to make a PR?

**Reply by KaKi87** [reply]
Alright.
